stat.go

```diff
@@ -648,6 +648,94 @@ func KullbackLeibler(p, q []float64) float64 {
 	return kl
 }
 
+// LinearRegression computes the best-fit line
+//  y = alpha*x + beta
+// to the data in x and y with the given weights. If origin is true, the
+// regression is forced to pass through the origin.
+//
+// Specifically, LinearRegression computes the values of alpha and
+// beta such that the total residual
+//  \sum_i w[i]*(y[i] - alpha*x[i] - beta)^2
+// is minimized. If origin is true, then beta is forced to be zero.
+//
+// The lengths of x and y must be equal. If weights is nil then all of the
+// weights are 1. If weights is not nil, then len(x) must equal len(weights).
+func LinearRegression(x, y, weights []float64, origin bool) (alpha, beta float64) {
+	if len(x) != len(y) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(weights) != len(x) {
+		panic("stat: slice length mismatch")
+	}
+
+	w := 1.0
+	if origin {
+		var x2Sum, xySum float64
+		for i, xi := range x {
+			if weights != nil {
+				w = weights[i]
+			}
+			yi := y[i]
+			xySum += w * xi * yi
+			x2Sum += w * xi * xi
+		}
+		beta = xySum / x2Sum
+
+		return alpha, beta
+	}
+
+	beta = Covariance(x, y, weights) / Variance(x, weights)
+	alpha = Mean(y, weights) - beta*Mean(x, weights)
+
+	return alpha, beta
+}
+
+// RSquared returns the coefficient of determination for the best-fit
+// line
+//  y = alpha*x + beta
+// and the data in x and y with the given weights.
+//
+// The lengths of x and y must be equal. If weights is nil then all of the
+// weights are 1. If weights is not nil, then len(x) must equal len(weights).
+func RSquared(x, y, weights []float64, alpha, beta float64) float64 {
+	if len(x) != len(y) {
+		panic("stat: slice length mismatch")
+	}
+	if weights != nil && len(weights) != len(x) {
+		panic("stat: slice length mismatch")
+	}
+
+	w := 1.0
+	if alpha == 0 {
+		var ssr, tot float64
+		for i, xi := range x {
+			if weights != nil {
+				w = weights[i]
+			}
+			fi := beta * xi
```

Comment:
Sorry, I didn't realize R^2 was so messed up.

I'm having a tough time removing my statistician hat, so this is a long reply.

In the article you link `Heedlessly applying equation (4) would then result in an implausibly negative (and thus uninterpretable coefficient of determination`. I'm sorry, but I think that's wrong. R^2 is _not_ a goodness of fit metric, at least not directly. It's a measure of variance reduction, or from Wikipedia `the proportion of the variance in the dependent variable that is predictable from the independent variable`. One can have bad models with high R^2, and the correct model with low R^2 (see section 2.1.1 of [Cosma Shalizi's notes](http://www.stat.cmu.edu/~cshalizi/402/lectures/02-truth-about-linear-regression/lecture-02.pdf) ). Since R^2 is a measure of the reduced variance, a negative R^2 has a very clear interpretation -- the fit has actually increased the variance of the data!

There does appear to be debate on the actual definition of R^2. Some define it as `1 - SSres / SStot` (like Matlab), and others define it as `SSreg / SStot`, (like Minitab). It seems clear to me that we should use the first definition (which is eq. 4 in your link).  This definition works for all lines (and really all regressors, linear or not). If it happens to be that `cov(f_i, y_i) = 0` for the regression, (as is true for OLS) then the other definition is a neat fact. If we take the second equation as a definition, then R^2 _only_ makes sense for OLS with an intercept. 

Let's take an extreme example, and consider the predictor `f_i = inf`. With the first definition, `R^2 = -inf`; that's clearly really bad. Under the second definition, R^2 is `inf`. Is that really good? Similarly, there exists a `f(x) = const.` line where `SSreg = SStot`. Thus, I get an R^2 of 1 even though I clearly have a bad predictor. The same flaw does not exist under the residual definition. The only way to have an R^2 of 1 is to make perfect predictions everywhere. (The same flaw, by the way, exists with the definition of 4' in your article).

In short, I think we should define R^2 based on the residuals. The equation is simple, and it's clear what's being computed. It computes something reasonable for all lines, not just the single best-fit line. Finally, the answer is the same in the common case of finding R^2 to the best fit line. Since there does seem to be disagreement on the "true" definition, I feel that we are free to do the reasonable one as long as we're clear about it.